ftw/simplelayout/browser/provider.py

```diff
@@ -33,37 +33,40 @@ def _prepare_html(self, text):
         return re.sub(">\s*<", "><", text), type_
 
 
-class BaseSimplelayoutExpression(object):
+class SimplelayoutRenderer(object):
 
+    layout_template = CleanedViewPageTemplateFile('templates/layout.pt')
+    slot_template = CleanedViewPageTemplateFile('templates/slot.pt')
     fallbackview = ViewPageTemplateFile('templates/render_block_error.pt')
-    structure = CleanedViewPageTemplateFile('templates/structure.pt')
 
-    @property
-    def one_layout_one_column(self):
-        return [{"cols": [{"blocks": []}]}]
+    def __init__(self, context, storage, name, view=None):
 
-    def user_can_edit(self):
-        return api.user.has_permission(
-            'Modify portal content',
-            obj=self.context) and int(self.request.get('disable_border', 0)) == 0
+        self.context = context
+        self.request = self.context.REQUEST
+        self.storage = storage
+        self.name = name
+        self.view = view
 
-    def get_css_class(self):
-        css_classes = ['sl-simplelayout']
-        if self.user_can_edit():
-            css_classes.append("sl-can-edit")
-        return " ".join(css_classes)
+    def render_slot(self):
+        return self.slot_template()
 
-    def rows(self):
-        """ Return datastructure for rendering blocks.
-        """
-        page_conf = IPageConfiguration(self.context)
-        blocks = self._blocks()
+    def render_layout(self, index=None):
+
+        self.rows = self.storage.get(self.name, self._one_layout_one_column())
```

Comment:
I think this should be a local variable:
```diff
-         self.rows = self.storage.get(self.name, self._one_layout_one_column())
+         rows = self.storage.get(self.name, self._one_layout_one_column())
```

- it is only every used within `render_layout`
- it may be modified within `render_layout`, depending on `index`, and thus should not be exposed as its state is a side effect.